adminapi/base.py

```diff
@@ -0,0 +1,450 @@
+from distutils.util import strtobool
+from itertools import chain
+
+
+class DatasetError(Exception):
+    pass
+
+
+class BaseQuerySet(object):
+    def __init__(self, filters):
+        self._filters = filters
+        self._results = None
+        self._restrict = None
+        self._order_by = None
+
+    def __iter__(self):
+        return iter(self.get_results())
+
+    def __len__(self):
+        return len(self.get_results())
+
+    def __bool__(self):
+        return bool(self.get_results())
+
+    def __repr__(self):
+        # QuerySet is not used directly but through query function
+        kwargs = ', '.join(
+            '{0}={1!r}'.format(k, v) for k, v in self._filters.items()
+        )
+        query_repr = 'query({0})'.format(kwargs)
+
+        if self._restrict:
+            query_repr += '.restrict({0})'.format(', '.join(self._restrict))
+
+        return query_repr
+
+    def get_lookup(self, attr):
+        lookup = {}
+        for host in self:
+            if attr in host:
+                lookup[host[attr]] = host
+        return lookup
+
+    def restrict(self, *attrs):
+        if not attrs:
+            return self
+
+        if isinstance(attrs[0], (list, set, tuple)):
+            attrs = attrs[0]
+
+        self._restrict = {str(a) for a in attrs}
+
+        return self
+
+    def get(self):
+        results = self.get_results()
+        if len(results) != 1:
+            raise DatasetError('get() requires exactly 1 matched object')
+        return results[0]
+
+    def is_dirty(self):
+        return any(s.is_dirty() for s in self)
+
+    def commit(self, skip_validation, force_changes):
+        raise NotImplementedError()
+
+    def rollback(self):
+        for obj in self:
+            obj.rollback()
+        return self
+
+    def delete(self):
+        for obj in self:
+            obj.delete()
+        return self
+
+    def update(self, **attrs):
+        for obj in self:
+            if not obj.is_deleted():
+                obj.update(attrs)
+        return self
+
+    def iterattrs(self, attr='hostname'):
```

Comment:
`iterattrs` would assume you can provide several attributes. Maybe `iterattr` would be better?